--- a/WebappConfig/permissions.py
+++ b/WebappConfig/permissions.py
@@ -179,6 +179,12 @@ class PermissionMap:
             self.__permissions = splitted_permissions
             self.__absolute    = False

+    def __str__(self):
+      if self.__absolute:
+        return 'Absolute: {}'.format(oct(self.__permissions))
+      else:
+        return 'Relative: {}'.format(self.__permissions)
+
 def get_group(group):
     '''
     Specify a group id either as integer, as string that can

--- a/WebappConfig/db.py
+++ b/WebappConfig/db.py
@@ -549,7 +549,7 @@ class WebappSource(AppHierarchy):
             flist = open(self.appdir() + '/' + server_owned)
             server_files = flist.readlines()

-            OUT.debug('Identified config-protected files.', 7)
+            OUT.debug('Identified server-owned files.', 7)

             flist.close()


