From 7cf53f3da6040428543bd254e7219ee3896be127 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Diego=20Elio=20Petten=C3=B2?= <flameeyes@flameeyes.eu>
Date: Fri, 17 Aug 2012 19:06:02 -0700
Subject: [PATCH 3/8] configure: use pg_config to find where to find
 PostgreSQL libraries.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit


Signed-off-by: Diego Elio Pettenò <flameeyes@flameeyes.eu>
---
 configure.in |   51 +++++++++++++++++++++++++++------------------------
 1 files changed, 27 insertions(+), 24 deletions(-)

diff --git a/configure.ac b/configure.ac
index 989bf8f..4257c6d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -213,27 +213,34 @@ AC_ARG_WITH(pgsql,
 	PGSQL=$withval,)
 AC_CHECK_LIB(crypt,main)
 if test "$ac_cv_lib_crypt_main" = "yes" -a "x$PGSQL" != "xno"; then
-  if test -n "$PGSQL"; then
-    LDFLAGS="$LDFLAGS -L$PGSQL/lib"
-    CPPFLAGS="$CPPFLAGS -I$PGSQL/include"
+  if test -x $with_pgsql/bin/pg_config ; then
+    np_pg_config="$with_pgsql/bin/pg_config"
   fi
-  AC_CHECK_LIB(pq,PQsetdbLogin,,,-lcrypt)
-  if test "$ac_cv_lib_pq_PQsetdbLogin" = "yes"; then
-    AC_CHECK_HEADERS(pgsql/libpq-fe.h)
-    AC_CHECK_HEADERS(postgresql/libpq-fe.h)
-    AC_CHECK_HEADERS(libpq-fe.h)
-    if [[ -n "$PGSQL" -a "$ac_cv_header_libpq_fe_h" = "yes" ]]; then
-      PGLIBS="-L$PGSQL/lib -lpq -lcrypt"
-      PGINCLUDE="-I$PGSQL/include"
-    elif test  "$ac_cv_header_pgsql_libpq_fe_h" = "yes"; then
-      PGLIBS="-lpq -lcrypt"
-      PGINCLUDE="-I/usr/include/pgsql"
-    elif test  "$ac_cv_header_postgresql_libpq_fe_h" = "yes"; then
-      PGLIBS="-L$PGSQL/lib -lpq -lcrypt"
-      PGINCLUDE="-I/usr/include/postgresql"
-    elif test  "$ac_cv_header_libpq_fe_h" = "yes"; then
-      PGLIBS="-L$PGSQL/lib -lpq -lcrypt"
-      PGINCLUDE="-I$PGSQL/include"
+  if test -z "$np_pg_config"; then
+    with_pgsql="no"
+  else
+    if test -n "$PGSQL"; then
+      LDFLAGS="$LDFLAGS -L`$np_pg_config --libdir`"
+      CPPFLAGS="$CPPFLAGS -I`$np_pg_config --includedir`"
+    fi
+    AC_CHECK_LIB(pq,PQsetdbLogin,,,-lcrypt)
+    if test "$ac_cv_lib_pq_PQsetdbLogin" = "yes"; then
+      AC_CHECK_HEADERS(pgsql/libpq-fe.h)
+      AC_CHECK_HEADERS(postgresql/libpq-fe.h)
+      AC_CHECK_HEADERS(libpq-fe.h)
+      if [[ -n "$PGSQL" -a "$ac_cv_header_libpq_fe_h" = "yes" ]]; then
+        PGLIBS="-L`$np_pg_config --libdir` -lpq -lcrypt"
+        PGINCLUDE="-I`$np_pg_config --includedir`"
+      elif test  "$ac_cv_header_pgsql_libpq_fe_h" = "yes"; then
+        PGLIBS="-lpq -lcrypt"
+        PGINCLUDE="-I`$np_pg_config --includedir`"
+      elif test  "$ac_cv_header_postgresql_libpq_fe_h" = "yes"; then
+        PGLIBS="-L`$np_pg_config --libdir` -lpq -lcrypt"
+        PGINCLUDE="-I`$np_pg_config --includedir`"
+      elif test  "$ac_cv_header_libpq_fe_h" = "yes"; then
+        PGLIBS="-L`$np_pg_config --libdir` -lpq -lcrypt"
+        PGINCLUDE="-I`$np_pg_config --includedir`"
+      fi
     fi
     if test -z "$PGINCLUDE"; then
       AC_MSG_WARN([Skipping PostgreSQL plugin (check_pgsql)])
@@ -243,10 +250,6 @@ if test "$ac_cv_lib_crypt_main" = "yes" -a "x$PGSQL" != "xno"; then
       AC_SUBST(PGINCLUDE)
       EXTRAS="$EXTRAS check_pgsql"
     fi
-  else
-    AC_MSG_WARN([Skipping PostgreSQL plugin (check_pgsql)])
-    AC_MSG_WARN([LIBS="$LIBS" CPPFLAGS="$CPPFLAGS"])
-    AC_MSG_WARN([install PostgreSQL libs to compile this plugin (see REQUIREMENTS).])
   fi
 else
   AC_MSG_WARN([Skipping PostgreSQL plugin (check_pgsql)])
-- 
1.7.8.6

