diff -u -Naur a/configure.ac b/configure.ac
--- a/configure.ac	2014-08-10 16:41:27.684199231 +0200
+++ b/configure.ac	2014-08-10 16:41:52.737949777 +0200
@@ -1737,7 +1737,7 @@
 AC_SUBST(DEPLIBS)
 
 AM_GNU_GETTEXT([external], [need-ngettext])
-AM_GNU_GETTEXT_VERSION(0.15)
+AM_GNU_GETTEXT_VERSION(0.18.3)
 
 dnl Check for Redhat spopen problem
 dnl Wierd problem where ECHILD is returned from a wait call in error
